Support kernel ELF compressed files

atenart opened the pull request:
One thing to note is the lack of automated tests. Of course this should be fixed but is not trivial:

- We don't want to include full kernel images (size issue).
- We should add a way to generate the full `tests/data` content as it's getting more and more complex to get right.

Based on #18.

amorenoz commented:
> One thing to note is the lack of automated tests. Of course this should be fixed but is not trivial:
> 
>     We don't want to include full kernel images (size issue).

Maybe we could craft a kernel config with the bare minimum and build a small image and 2 or 3 modules.
Also, we could store it in an independent repo (submodule?).

>     We should add a way to generate the full tests/data content as it's getting more and more complex to get right.

atenart commented on src/utils/elf.rs:
```diff
@@ -77,3 +82,101 @@ pub fn collection_from_kernel_dir<P: AsRef<Path>>(path: P) -> Result<BtfCollecti
 }
+
+#[cfg(feature = "elf-compressed")]
+mod compression {
```
My goal was to put it under `elf::compression` and that was easier this way w/o moving the elf part to its own directory (as the compression part isn't that big. But I have no strong feeling here, I can move them if you prefer.

atenart commented on .cirrus.yml:
```diff
@@ -18,9 +18,11 @@ test_task:
   check_script:
     - cargo fmt --check
-    - cargo clippy -F elf,regex -- -D warnings
+    - cargo clippy -F elf,elf-compressed,regex -- -D warnings
```
That's right, I just like to keep things explicit (here: enabling all features; as opposed to the test matrix above where it's not explicit). I can remove if you prefer though.

amorenoz replied:
In that case we should add it to all the other `cargo build` lines in this file, right?

atenart commented:
> > One thing to note is the lack of automated tests. Of course this should be fixed but is not trivial:
> > ```
> > We don't want to include full kernel images (size issue).
> > ```
> 
> Maybe we could craft a kernel config with the bare minimum and build a small image and 2 or 3 modules.

That's an option. One set for each compression alg.

> Also, we could store it in an independent repo (submodule?).

I think that's fine, those are excluded from crates.io anyway.

vlrpl commented on Cargo.toml:
```diff
@@ -18,12 +18,17 @@ maintenance = { status = "actively-developed" }
 [dependencies]
 anyhow = "1.0"
 byteorder = "1.5"
+bzip2 = { version = "0.4", optional = true }
```
I assume packaged versions are compatible with this one, right?

`Version        : 0.5.2`

vlrpl reviewed (commented):
IMO, it's worth it moving this forward to a non-RFC PR.

vlrpl commented:
> config



> > One thing to note is the lack of automated tests. Of course this should be fixed but is not trivial:
> > ```
> > We don't want to include full kernel images (size issue).
> > ```
> 
> Maybe we could craft a kernel config with the bare minimum and build a small image and 2 or 3 modules. Also, we could store it in an independent repo (submodule?).
> 

As a small config I suggest to use `tools/testing/selftests/net/config`. Should be good enough for what we need.

> > ```
> > We should add a way to generate the full tests/data content as it's getting more and more complex to get right.
> > ```

atenart commented on src/utils/elf.rs:
```diff
@@ -77,3 +82,101 @@ pub fn collection_from_kernel_dir<P: AsRef<Path>>(path: P) -> Result<BtfCollecti
+    }
+
+    pub(super) fn try_decompress(file: Vec<u8>) -> Result<Vec<u8>> {
+        for i in 0..file.len() {
```
The try and see logic is for the specific image format of a kernel image. But this also works on more classic ELF files that were just compressed, such as kernel modules (which should work on the first iteration).

atenart commented:
This PR is now ready. Major changes:

- Dropped lzma support, as this is more complex to handle and my understanding is it is more a legacy format and people should use xz+lzma2 anyway. This in line with the compression algs supported by kernel modules too.
- Added a script to generate all the test assets, supporting all the compression algs introduced here too.
- Added tests for the compression algs introduced here.

atenart commented on Cargo.toml:
```diff
@@ -17,13 +17,18 @@ maintenance = { status = "actively-developed" }
 byteorder = "1.5"
+bzip2 = { version = "0.6", optional = true }
 elf = { version = "0.7", optional = true }
+flate2 = { version = "1.0", optional = true }
```
Sure, we can use the latest release.
